### ISSUE TITLE
NMS Very Slow when Large Input Size

### ISSUE THREAD

**Issue by LeoSouquet** [ultralytics/yolov5#6639]
### Search before asking

- [X] I have searched the YOLOv5 [issues](https://github.com/ultralytics/yolov5/issues) and [discussions](https://github.com/ultralytics/yolov5/discussions) and found no similar questions.

### Question

Hi All,

I am trying to understand why is NMS so slow in my current inference.
My input size is very large, 1920*800,  I have therefore a lot of predictions from the model.

After lots of tests, the line slowing everything down is in NMS : _**x = x[xc[xi]]**_

This line seems to do the computation on the CPU.
I say that because this line takes approximately 30ms in my case.
If I transfer the predictions before NMS onto the CPU, the transfer takes 30ms but this line will only take 3ms.

That is why I deduced that this line seems to be executed on the CPU.

Any idea why? and also how to let the computation be done on CPU?
Thanks in advance,

Regards,

`

def non_max_suppression(prediction, conf_thres=0.25, iou_thres=0.45, classes=None, agnostic=False, multi_label=False,
                        labels=(), max_det=300):

    nc = prediction.shape[2] - 5  # number of classes
    xc = prediction[..., 4] > conf_thres  # candidates

    [...] 
    output = [torch.zeros((0, 6), device=prediction.device)] * prediction.shape[0]
    for xi, x in enumerate(prediction):  # image index, image inference
        # Apply constraints
        x[((x[..., 2:4] < min_wh) | (x[..., 2:4] > max_wh)).any(1), 4] = 0  # width-height
        x = x[xc[xi]]  <- **This line seem to compute on CPU**
`

**Comment by glenn-jocher** [comment]
@LeoCyclope 👋 hi, thanks for letting us know about this possible problem with YOLOv5 🚀. I'm unable to produce slow NMS speeds as you describe, even at image size 1920 (I even used Colab free here, which is about the slowest backend possible):
<img width="1498" alt="Screenshot 2022-02-14 at 15 46 07" src="https://user-images.githubusercontent.com/26833433/153885940-84d29f56-c292-4de0-b2e1-11d2e41f6ea6.png">

We've created a few short guidelines below to help users provide what we need in order to get started investigating a possible problem.

## How to create a Minimal, Reproducible Example

When asking a question, people will be better able to provide help if you provide **code** that they can easily understand and use to **reproduce** the problem. This is referred to by community members as creating a [minimum reproducible example](https://docs.ultralytics.com/help/minimum_reproducible_example/). Your code that reproduces the problem should be:

- ✅ **Minimal** – Use as little code as possible to produce the problem
- ✅ **Complete** – Provide all parts someone else needs to reproduce the problem
- ✅ **Reproducible** – Test the code you're about to provide to make sure it reproduces the problem

For [Ultralytics](https://ultralytics.com/) to provide assistance your code should also be:

- ✅ **Current** – Verify that your code is up-to-date with GitHub [master](https://github.com/ultralytics/yolov5/tree/master), and if necessary `git pull` or `git clone` a new copy to ensure your problem has not already been solved in [master](https://github.com/ultralytics/yolov5/tree/master).
- ✅ **Unmodified** – Your problem must be reproducible using official YOLOv5 code without changes. [Ultralytics](https://ultralytics.com/) does not provide support for custom code ⚠️.

If you believe your problem meets all the above criteria, please close this issue and raise a new one using the 🐛 **Bug Report** [template](https://github.com/ultralytics/yolov5/issues/new/choose) with a [minimum reproducible example](https://docs.ultralytics.com/help/minimum_reproducible_example/) to help us better understand and diagnose your problem. 

Thank you! 😃

**Comment by LeoSouquet** [comment]
Hi Glenn,

Thanks so much :)
Indeed, using the val.py, the NMS is no longer slow.
This comes from the fact that in the val.py the transfert GPU to CPU is done somewhere else than the code I tested.

What I don't understand is why is transferring the image from CPU to GPU significantly faster than transferring the predictions at the end of the inference from the GPU to the CPU?

Any intuition behind that?

Thanks a lot!

**Comment by DensenDavis** [comment]
Hi, although stale, I am writing this since I ran into the same issue.

Why the line : x = x[xc[xi]] in nms seems slow is probably because you might have used time.time() plainly before and after this operation to calculate the delay in the execution of this line. Since the variables involved in this operation are in gpu, you need to add torch.cuda.synchronize() right before you call time.time(). After this change, you will notice that the nms time is negligible compared to the model prediction time.

**Comment by LeoSouquet** [comment]
Hi DensenDavis !!

Thank you soooo much for this! It is much much apprecaited it :)
I will test this when I have the change but thanks a lot :)

Wish you a happy new year :)

**Comment by glenn-jocher** [comment]
@LeoSouquet you're welcome! I'm glad to be of help 😊. Wishing you a fantastic new year as well! If you have any more questions or need further assistance, feel free to ask. Happy coding and best wishes!
